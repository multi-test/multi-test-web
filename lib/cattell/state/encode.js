import { BitStream } from 'bit-buffer';
import { toBase64 } from "./base64";
function answerToBit2(answer) {
    switch (answer) {
        case '': return 0b00;
        case 'A': return 0b01;
        case 'B': return 0b10;
        case 'C': return 0b11;
    }
}
function genderToBit2(gender) {
    switch (gender) {
        case '': return 0b00;
        case 'M': return 0b01;
        case 'F': return 0b10;
    }
}
export function encodeState(state) {
    const bitStream = new BitStream(new ArrayBuffer(49));
    {
        const byte = state.position;
        bitStream.writeUint8(byte);
    }
    for (const answer of state.answers) {
        const bit2 = answerToBit2(answer);
        bitStream.writeBits(bit2, 2);
    }
    {
        const bit2 = genderToBit2(state.profile.gender);
        bitStream.writeBits(bit2, 2);
    }
    {
        const bit6 = state.profile.age === 0 ? 0 : state.profile.age - 15;
        bitStream.writeBits(bit6, 6);
    }
    bitStream.index = 0;
    let base64Data = '';
    for (let i = 0; i < 65; i++) {
        base64Data += toBase64(bitStream.readBits(6, false));
    }
    const base64Name = btoa(unescape(encodeURIComponent(state.profile.name)));
    return base64Data + base64Name;
}
