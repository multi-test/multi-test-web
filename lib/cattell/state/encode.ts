import {CattellState} from "./type";
import {WritableBitStream} from "./BitStream";

function answerToBit2(answer: '' | 'A' | 'B' | 'C'): 0 | 1 | 2 | 3 {
    switch (answer) {
        case '':  return 0b00;
        case 'A': return 0b01;
        case 'B': return 0b10;
        case 'C': return 0b11;
    }
}

function genderToBit2(gender: 'M' | 'F' | ''): 0 | 1 | 2 {
    switch (gender) {
        case '':  return 0b00;
        case 'M': return 0b01;
        case 'F': return 0b10;
    }
}

export function encodeState(state: CattellState): string {
    const bitStream = new WritableBitStream();
    bitStream.writeByte(state.position);

    for (const answer of state.answers) {
        bitStream.writeBit2(answerToBit2(answer));
    }

    bitStream.writeBit2(genderToBit2(state.profile.gender));
    bitStream.writeByte(state.profile.age);

    let hash = '';
    const bytes = bitStream.getBytes();

    // istanbul ignore next
    if (bytes !== null) {
        for (const byte of bytes) {
            hash += String.fromCharCode(byte);
        }
    }

    hash += state.profile.name;

    return btoa(unescape(encodeURIComponent(hash)))
}