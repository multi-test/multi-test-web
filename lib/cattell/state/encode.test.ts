import {encodeState} from './encode';
import createBlankCattellState from './blank';
import {CattellState} from "./type";

describe('cattell.state.blank', () => {
    let state: CattellState;

    beforeEach(() => {
        state = createBlankCattellState();
    });

    afterEach(() => {
        expect(encodeState(state)).toMatchSnapshot();
    });

    it('should create blank valid state matching the snapshot', () => {});

    it('should encode age', () => {
        state.profile.age = 70;
    });

    it('should encode name', () => {
        state.profile.name = 'Test '.repeat(12);
    });

    it('should encode male gender', () => {
        state.profile.gender = 'M';
    });

    it('should encode female gender', () => {
        state.profile.gender = 'F';
    });

    it('should encode A answers', () => {
        state.answers.fill('A');
    });

    it('should encode B answers', () => {
        state.answers.fill('B');
    });

    it('should encode C answers', () => {
        state.answers.fill('C');
    });
});

