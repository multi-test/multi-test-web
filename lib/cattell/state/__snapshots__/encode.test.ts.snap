// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cattell.state.blank should create blank valid state matching the snapshot 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=="`;

exports[`cattell.state.blank should encode A answers 1`] = `"AFVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVUVAA=="`;

exports[`cattell.state.blank should encode B answers 1`] = `"AMKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqwqrCqsKqKgA="`;

exports[`cattell.state.blank should encode C answers 1`] = `"AMO/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/w7/Dv8O/PwA="`;

exports[`cattell.state.blank should encode age 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAARg=="`;

exports[`cattell.state.blank should encode female gender 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADCgAA="`;

exports[`cattell.state.blank should encode male gender 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAA=="`;

exports[`cattell.state.blank should encode name 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAFRlc3QgVGVzdCBUZXN0IFRlc3QgVGVzdCBUZXN0IFRlc3QgVGVzdCBUZXN0IFRlc3QgVGVzdCBUZXN0IA=="`;
